### PR TITLE
Dynamic changes

### PR DESCRIPTION
**Pull request by ux-AylonS**

### CODE REVIEW

File: src/router/index.ts [hunk 1]
```diff
@@ -107,6 +108,20 @@ const routes: Array<RouteRecordRaw> = [
...
             ],
           },
+          {
+            path: 'dynamic-rules',
+            name: 'DynamicRules',
+            redirect: (route) => {
+              return `/${route.params.branch}/dynamic-rules/list`
+            },
+            children: [
+              {
```
Comment (Aviv-Galmidi) [inline src/router/index.ts:118]: We're missing the second child for the config

File: src/views/DynamicRulesList.vue [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+          <p class="control">
+            <button class="button is-small download-doc-button"
+                    :class="{'is-loading':isDownloadLoading}"
+                    @click="downloadDoc()"
+                    title="Download document"
+                    data-qa="download-document">
+                <span class="icon is-small">
+                    <i class="fas fa-download"></i>
+                </span>
+            </button>
```
Comment (Aviv-Galmidi) [inline src/views/DynamicRulesList.vue:15]: Changed from `main` branch:

```suggestion
        <div class="field is-grouped is-pulled-right">
          <p class="control">
            <button class="button is-small download-doc-button"
                    :class="{'is-loading':isDownloadLoading}"
                    @click="downloadDoc()"
                    title="Download document"
                    data-qa="download-document">
              <span class="icon is-small">
                <i class="fas fa-download"></i>
              </span>
              <span>
                Download
              </span>
            </button>
```

File: src/views/DynamicRulesList.vue [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    },
+
+    editDynamicRule(id: string) {
+      const routeToDoc = `/${this.selectedBranch}/dynamic-rules/config/${id}`
+      this.$router.push(routeToDoc)
+    },
+
+    downloadDoc() {
+      if (!this.isDownloadLoading) {
+        // const completeDoc = {this.dynamicRulesData + this.globalFiltersData}
```
Comment (Aviv-Galmidi) [inline src/views/DynamicRulesList.vue:204]: Let's remove this comment

File: src/views/DynamicRulesList.vue [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+          this.dynamicRulesData = []
+          this.isDownloadLoading = false
+        },
+      })
+      this.dynamicRulesData = response?.data || []
+
+      // bring Tags from GlobalFilters
+      this.globalFiltersData = []
+      this.dynamicRulesDataAndGlobalFilters = []
+      // const config = {headers: {'x-fields': 'id, tags, action'}} returns all fields
```
Comment (Aviv-Galmidi) [inline src/views/DynamicRulesList.vue:227]: I think it was commented out by mistake, we can use it in the /globalfilters/ request

File: src/views/DynamicRulesList.vue [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+            action: responseGlobal.data.action})
+          const joinedObject = JSON.parse(JSON.stringify(_.merge(doc, {tags: responseGlobal.data.tags,
+            action: responseGlobal.data.action}))) as DynamicRuleAndGlobalFilters
+          this.dynamicRulesDataAndGlobalFilters.push(joinedObject)
+        })
+      })
+      this.isDownloadLoading = false
+    },
+
+    async switchBranch() {
```
Comment (Aviv-Galmidi) [inline src/views/DynamicRulesList.vue:246]: Method can be removed

File: src/views/DynamicRulesList.vue [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+      this.setLoadingDocStatus(true)
+      Utils.toast(`Switched to branch '${this.selectedBranch}'.`, 'is-info')
+      await this.loadDynamicRulesData()
+      this.setLoadingDocStatus(false)
+    },
+
+  },
+  async created() {
+    await this.branchesStore.list
```
Comment (Aviv-Galmidi) [inline src/views/DynamicRulesList.vue:255]:
```suggestion
    this.setLoadingDocStatus(true)
    await this.branchesStore.list
    this.setLoadingDocStatus(false)
```

File: src/views/DynamicRulesList.vue [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    await this.loadDynamicRulesData()
```
Comment (Aviv-Galmidi) [inline src/views/DynamicRulesList.vue:256]: This will be called from the watch on `selectedBranch`

File: src/views/DynamicRulesList.vue [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+          console.log('Error while attempting to load documents')
+          this.dynamicRulesDataAndGlobalFilters = []
+          this.isDownloadLoading = false
+        },
+        }).then((responseGlobal: AxiosResponse<GlobalFilter>) => {
+          this.globalFiltersData.push({id: responseGlobal.data.id, tags: responseGlobal.data.tags,
+            action: responseGlobal.data.action})
+          const joinedObject = JSON.parse(JSON.stringify(_.merge(doc, {tags: responseGlobal.data.tags,
+            action: responseGlobal.data.action}))) as DynamicRuleAndGlobalFilters
+          this.dynamicRulesDataAndGlobalFilters.push(joinedObject)
```
Comment (Aviv-Galmidi) [inline src/views/DynamicRulesList.vue:240]: This type seems confusing as the data is not really on the same object in the reality (Backend, APIs, etc.)

Why not use `dispalyFunction` like I suggested 

> the list columns will get the data by searching the Global Filters for the id with matching id to the row item

So `Tags` column would look something like
```
    {
      title: 'Tags',
      displayFunction: (item: DynamicRule) => {
        const matchingGlobalFilter = this.globalFiltersData.find((globalFilter: GlobalFilter) => {
         return globalFilter.id === `dr_${item.id}`
        }
        return matchingGlobalFilter.tags?.join('\n')
      },
      isSortable: false,
      isSearchable: true,
      classes: 'width-100px white-space-pre ellipsis',
    },
```
Note: I wrote this code in the GitHub comment section, so make sure it's actually okay in the system

**Review by Aviv-Galmidi — CHANGES_REQUESTED** [review]

File: src/router/index.ts (hunk 1 above)
Comment (ux-AylonS) [reply src/router/index.ts:118]: I use DocumentEditor to serve the DynamicRulesEditor,
and it is working well.

File: src/views/DynamicRulesList.vue (hunk 4 above)
Comment (ux-AylonS) [reply src/views/DynamicRulesList.vue:227]: I tried to use it but I still got back the full object and not only the listed columns

File: src/views/DynamicRulesList.vue [hunk 9]
```diff
@@ -223,37 +241,29 @@ export default defineComponent({
...
+        const config = {headers: {'x-fields': 'id, tags, action'}}
+        RequestsUtils.sendRequest({methodName: 'GET', url, config, onFail: () => {
           console.log('Error while attempting to load documents')
-          this.dynamicRulesDataAndGlobalFilters = []
           this.isDownloadLoading = false
         },
         }).then((responseGlobal: AxiosResponse<GlobalFilter>) => {
+          console.log('responseGlobal.data.id', responseGlobal.data.id,
+            'responseGlobal.data.tags', responseGlobal.data.tags)
           this.globalFiltersData.push({id: responseGlobal.data.id, tags: responseGlobal.data.tags,
```
Comment (Aviv-Galmidi) [inline src/views/DynamicRulesList.vue:255]: We can se this.globalFiltersData as responseGlobal.data

File: src/views/DynamicRulesList.vue (hunk 8 above)
Comment (ux-AylonS) [reply src/views/DynamicRulesList.vue]: DONE

File: src/views/DynamicRulesList.vue [hunk 10]
```diff
@@ -122,10 +126,14 @@ export default defineComponent({
...
           displayFunction: (item: DynamicRule) => {
-            const matchingGlobalFilter = _.find(this.globalFiltersData, (globalFilter: GlobalFilter) => {
-              return globalFilter.id === `dr_${item.id}`
-            })
-            return matchingGlobalFilter.tags?.join('\n')
+            if (this.globalFiltersData.length > 0) {
+              const matchingGlobalFilter = _.find(this.globalFiltersData, (globalFilter: GlobalFilter) => {
+                return globalFilter.id === `dr_${item.id}`
+              })
+              return matchingGlobalFilter ? matchingGlobalFilter.tags?.join('\n') : null
```
Comment (Aviv-Galmidi) [inline src/views/DynamicRulesList.vue:133]: displayFunction should return a string, let's return an empty string if the tags don't exist or there is no global filters data

File: src/views/DynamicRulesList.vue [hunk 11]
```diff
@@ -130,7 +130,7 @@ export default defineComponent({
               const matchingGlobalFilter = _.find(this.globalFiltersData, (globalFilter: GlobalFilter) => {
                 return globalFilter.id === `dr_${item.id}`
               })
-              return matchingGlobalFilter ? matchingGlobalFilter.tags?.join('\n') : null
+              return matchingGlobalFilter ? matchingGlobalFilter.tags?.join('\n') : ''
             } else {
               return []
```
Comment (Aviv-Galmidi) [inline src/views/DynamicRulesList.vue:135]: This should also return an empty string

**Review by Aviv-Galmidi — CHANGES_REQUESTED** [review]

File: src/views/DynamicRulesList.vue (hunk 11 above)
Comment (Aviv-Galmidi) [reply src/views/DynamicRulesList.vue:135]: Still not fixed

File: src/views/documentListConst.ts [hunk 12]
```diff
@@ -494,5 +494,35 @@ export const COLUMN_OPTIONS_MAP: ColumnOptionsMap = {
       classes: 'width-150px',
     },
   ],
+  'dynamic-rules': [
```
Comment (Aviv-Galmidi) [inline src/views/documentListConst.ts:497]: Did you re-add this by mistake

Comment (ux-AylonS) [reply src/views/documentListConst.ts]: yes, deleted

**Review by Aviv-Galmidi — APPROVED** [review]

**Review by Aviv-Galmidi — APPROVED** [review]

**Pull request merged**